### PR TITLE
[Compiler-v2] Fix issue 12255

### LINKED ISSUE #12255: [Bug][Compiler-v2] does not have the `copy` ability
# 🐛 Bug

When compiling the following code:

```
module 0x42::M {

    struct R has key { f: u64 }
    public fun f(r: R): (R, u64) {
        (r, 0)
    }
    public fun g(s: &signer) acquires R {
        let r = R{f:1};
        let i =3;
        (r,i) = f(r);
        move_to<R>(s, r);
    }
}
```

The compiler will generate the following error: 

```
error: local `r` of type `M::R` does not have the `copy` ability
   ┌─ tests/ability-check/resources.move:12:18
   │
12 │         (r, i) = f(r);
   │                  ^^^^ copy needed here because value is still in use
13 │         move_to<R>(s, r);
   │         ---------------- used here
```

### PR DESCRIPTION
**Pull request by rahxephon89**
### Description

This PR closes #12255 by adding a check for whether a variable is alive after a given program point, consider that this variable may be just introduced at this point. So a variable is alive if (a) it is in the after set of livevar (b) it is not in the instrs.dests() of the current instructions.

### Test Plan

1) existing tests pass
2) add a test `copy_ability_tuple.move`

### CODE REVIEW

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs [hunk 1]
```diff
@@ -172,6 +172,7 @@ impl<'a> TransferFunctions for CopyDropAnalysis<'a> {
         let temp_needs_drop = |temp: &TempIndex| {
             self.target.get_local_type(*temp).is_reference() || exit_state.may_return()
         };
+        let live_srcs_in_dests = live_var.live_src_temps_in_dests(instr);
         match instr {
             Assign(_, _, src, AssignKind::Inferred) => {
                 if temp_needs_copy(src) {
```
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:178]: You can actually have identity assign (`$t1 = $t1`), so we need to do this here too.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs [hunk 2]
```diff
@@ -172,6 +172,7 @@ impl<'a> TransferFunctions for CopyDropAnalysis<'a> {
         let temp_needs_drop = |temp: &TempIndex| {
             self.target.get_local_type(*temp).is_reference() || exit_state.may_return()
         };
+        let live_srcs_in_dests = live_var.live_src_temps_in_dests(instr);
```
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:175]: Only need this for few instructions so don't need to eagerly compute this all the time. But I would do this differently anyway.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs [hunk 3]
```diff
@@ -196,7 +197,9 @@ impl<'a> TransferFunctions for CopyDropAnalysis<'a> {
                 // point, is again used in the argument list. Also, in difference to assign inference, we only need
                 // to copy the argument if its not primitive.
                 for (i, src) in srcs.iter().enumerate() {
-                    if (temp_needs_copy(src) || srcs[i + 1..].contains(src)) && type_needs_copy(src)
+                    if !live_srcs_in_dests.contains(src)
+                        && (temp_needs_copy(src) || srcs[i + 1..].contains(src))
```
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:201]: Please keep this code clean:

1. `temp_needs_copy` is the single definition which determines whether we need to copy or not. Logic should go there.
2. You really should have a a function `livevar.is_temp_used_after(temp, instr)` and use that in the definition of `temp_needs_copy`.
3. `fn is_temp_used_after(&self, instr) { self.after.contains_key(temp) && !instrs.dests().contains(temp) }`

No need to compute the live_srcs_in_dests.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs (hunk 1 above)
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs]: done

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs (hunk 2 above)
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs]: done

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs (hunk 3 above)
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs]: done

**Review by wrwg — APPROVED** [review]

File: third_party/move/move-compiler-v2/src/pipeline/livevar_analysis_processor.rs [hunk 4]
```diff
@@ -77,6 +77,11 @@ impl LiveVarInfoAtCodeOffset {
         result
     }
 
+    /// Check whether temp is used after bc
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/pipeline/livevar_analysis_processor.rs:80]:
```suggestion
    /// Check whether `temp` is alive after `bc`.
    /// If `temp` is re-defined at `bc`, then we consider it not to be alive (killed by the re-definition).
```

**Review by vineethk — APPROVED** [review]
Minor suggestion, LGTM otherwise.

**Pull request merged**